src/lib.rs

```diff
@@ -1,123 +1,441 @@
+
+    let root_ca = collateral
+        .tcb_info_and_qe_identity_issuer_chain
+        .last()
+        .context("tcb issuer chain is empty")?;
+
+    // Verify the root certificate is self issued
+    if root_ca.tbs_certificate.issuer != root_ca.tbs_certificate.subject {
+        bail!("root certificate is not self issued");
+    }
+
+    let spki = x509_cert::spki::SubjectPublicKeyInfo::<Any, _>::from_pem(INTEL_ROOT_CA_PEM)?;
+    let intel_root_ca = X509VerifyingKey::try_from(spki).unwrap();
+    intel_root_ca
+        .verify(root_ca)
+        .context("Root CA signature verification failed")?;
+
+    // Build initial trust store with the root certificate
+    let mut trust_store = TrustStore::new(current_time, vec![root_ca.clone()])?;
+
+    // Verify that the CRL is signed by Intel and add it to the store.
+    trust_store
+        .add_crl(collateral.root_ca_crl.clone(), true, None)
+        .context("failed to verify root ca crl")?;
+
+    trust_store
+        .add_crl(collateral.platform_ca_crl.clone(), false, None)
+        .context("failed to verify platform ca crl")?;
+
+    trust_store
+        .add_crl(collateral.processor_ca_crl.clone(), false, None)
+        .context("failed to verify processor ca crl")?;
 
-        let tcbinfov3_json = include_str!("../data/tcbinfov3_00806f050000.json");
-        let tcbinfov3: TcbInfoV3 = serde_json::from_str(tcbinfov3_json).unwrap();
-        let tcbinfov3_serialize = serde_json::to_string(&tcbinfov3).unwrap();
-        assert!(tcbinfov3_serialize == tcbinfov3_json);
+    // Verify PCK Cert Chain and add it to the store.
+    let pck_cert_chain = quote.signature.pck_cert_chain.clone();
+    trust_store
+        .verify_chain_leaf(&pck_cert_chain)
+        .context("failed to verify pck crl issuer chain")?;
 
-        let sgx_signing_cert_pem = &parse_pem(include_bytes!("../data/signing_cert.pem")).unwrap()[0];
-        let sgx_signing_cert = parse_x509_der(&sgx_signing_cert_pem.contents);
+    // Verify TCB Info Issuer Chain
+    let tcb_issuer = trust_store
+        .verify_chain_leaf(&collateral.tcb_info_and_qe_identity_issuer_chain)
+        .context("failed to verify tcb info issuer chain")?;
 
-        assert!(validate_tcbinfov3(&tcbinfov3, &sgx_signing_cert, PINNED_TIME));
+    // Get TCB Signer Public Key
+    let tcb_signer = tcb_issuer
+        .cert
+        .tbs_certificate
+        .subject_public_key_info
+        .subject_public_key
+        .as_bytes()
+        .context("missing tcb signer public key")?;
+
+    // We are making big assumption here that the key is ECDSA P-256
+    let tcb_signer = p256::ecdsa::VerifyingKey::from_sec1_bytes(tcb_signer)
+        .context("invalid tcb signer public key")?;
+
+    // Verify the TCB Info
+    let tcb_info = collateral
+        .tcb_info
+        .as_tcb_info_and_verify(tcb_signer)
+        .context("failed to verify tcb info signature")?;
+
+    // Verify the quote's pck signing certificate chain
+    let _pck_signer = trust_store
```

Comment:
Duplicate, since you have already verified PCK Chain Issuer in line 121

Reply:
Done.